### ISSUE TITLE
Including `compress/zlib/reader` causes other file reads to not compile

### ISSUE THREAD

**Issue by viraptor** [crystal-lang/crystal#12905]
## Bug Report

When trying to compile the following:
```
require "json"
require "compress/zlib/reader"

types = File.open(ARGV[0]) do |f|
  Hash(String, String).from_json(f.gets_to_end)
end
```

I get error:

```
Error: can't infer the type of instance variable '@flate_io' of Compress::Zlib::Reader

Could you add a type annotation like this

    class Compress::Zlib::Reader
      @flate_io : Type
    end

replacing `Type` with the expected type of `@flate_io`?
```

Full trace:
<details>
In test.cr:4:14

 4 | types = File.open(ARGV[0]) do |f|
                  ^---
Error: instantiating 'File.class#open(String)'

In test.cr:4:14

 4 | types = File.open(ARGV[0]) do |f|
                  ^---
Error: instantiating 'File.class#open(String)'

In test.cr:5:24

 5 | Hash(String, String).from_json(f.gets_to_end)
                          ^--------
Error: instantiating 'Hash(String, String).class#from_json(String)'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/from_json.cr:12:29

 12 | parser = JSON::PullParser.new(string_or_io)
                                ^--
Error: instantiating 'JSON::PullParser.class#new(String)'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/pull_parser.cr:109:5

 109 | next_token
       ^---------
Error: instantiating 'next_token()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/pull_parser.cr:649:12

 649 | @lexer.next_token
              ^---------
Error: instantiating 'JSON::Lexer+#next_token()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/lexer.cr:33:5

 33 | skip_whitespace
      ^--------------
Error: instantiating 'skip_whitespace()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/lexer.cr:85:7

 85 | next_char
      ^--------
Error: instantiating 'next_char()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/lexer.cr:314:5

 314 | next_char_no_column_increment
       ^----------------------------
Error: instantiating 'next_char_no_column_increment()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/json/lexer/io_based.cr:11:25

 11 | @current_char = @io.read_char || '\0'
                          ^--------
Error: instantiating 'IO+#read_char()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/io.cr:302:17

 302 | peek = self.peek unless decoder
                   ^---
Error: instantiating 'IO+#peek()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/io/buffered.cr:103:7

 103 | fill_buffer
       ^----------
Error: instantiating 'fill_buffer()'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/io/buffered.cr:261:12

 261 | size = unbuffered_read(Slice.new(in_buffer, @buffer_size)).to_i
              ^--------------
Error: instantiating 'unbuffered_read(Slice(UInt8))'

In /Users/viraptor/.asdf/installs/crystal/1.6.0/src/compress/zlib/reader.cr:67:18

 67 | read_bytes = @flate_io.read(slice)
                   ^--------
Error: can't infer the type of instance variable '@flate_io' of Compress::Zlib::Reader

Could you add a type annotation like this

    class Compress::Zlib::Reader
      @flate_io : Type
    end

replacing `Type` with the expected type of `@flate_io`?
</details>

This happens when using 1.6.2 on MacOS (intel)

**Comment by straight-shoota** [comment]
`compress/zlib/reader` hasn't all its dependencies declared. Without that the compiler doesn't know about `Compress::Deflate::Reader.new` and thus can't infer the type of the ivar:

https://github.com/crystal-lang/crystal/blob/16f839ba250ad96054a68788a64cb671973427d0/src/compress/zlib/reader.cr#L18

It works if you require `compress/zlib`.

**Comment by viraptor** [comment]
That's super confusing, because it doesn't kick in unless I use `File.open`. And when I do, I'm not even explicitly touching the `Zlib` parts.

Does it need a fix in https://crystal-lang.org/api/1.6.2/Compress/Zlib/Reader.html ? That page only mentions:
```
Defined in:
compress/zlib/reader.cr
```

**Comment by straight-shoota** [comment]
Yeah, the compiler only tries to type `@flate_io` because the other code creates a call to `IO#read`, which traverses into `Compress::ZLib::Reader#unbuffered_read` and there the ivar is used. Maybe that's already a bit weird because you never create an instance of that type.

The docs page is correct. That's where the type is defined.
However, this definition does not necessarily mean that you need to *only* require this file to work with it.

I think the general idea is that you should require the entire package `compress/zlib`, not individual components from it.
That's never explicitly mentioned anywhere. And it's hard to keep track of.

IMO it would be best to declare all dependencies in that file directly. We can do that here simply for this single case. But the same applies to many other files. There's no easy way to validate the result except for creating tiny programs for every single source file that you would require that way. That's not very practical.

Maybe the best thing we can do right now is add a usage example to the API docs with `require "compress/zlib"`.

**Comment by HertzDevil** [comment]
Related: #12936
